tf2onnx/graph_helper.py

```diff
@@ -0,0 +1,75 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT license.
+
+"""
+tf2onnx.graph_helper - class to help building graph, such as helping to make complex node
+"""
+
+import numpy as np
+from tf2onnx import utils
+
+
+class GraphHelper(object):
+    """help to build graph"""
+    def __init__(self, graph):
+        self._g = graph
+
+    @property
+    def g(self):
+        return self._g
+
+    def make_slice(self, inputs_map, name=None, op_name_scope=None,
+                   outputs=None, shapes=None, dtypes=None, domain=None):
+        """
+        slice changes its schema at opset 10: it treats some attributes as dynamic input
+        so this function has to process inputs_map according to graph's opset version
+        to get "inputs" and "attr" to feed "make_node"
+        """
+        if self.g.opset < 10:
+            # "data" is string
+            # "starts", "ends" and "axes" are attributes, and "axes" is optional.
+            attr = {}
+            inputs = [inputs_map["data"]]
+            for key, required in {"starts": True, "ends": True, "axes": False}.items():
+                is_optional = not required
+                val = self._process_slice_input(inputs_map, key, is_optional)
+                if val is not None:
+                    attr[key] = val
+        else:
+            # slice-10 has 3 required inputs "data", "starts", "ends"l
+            # and 2 optional inputs "axes", "steps"
+            # input sequence should be "data", "starts", "ends", "axes", "steps"
+            attr = None
+            required_inputs = ["data", "starts", "ends"]
+            data, starts, ends = [self._process_slice_input(inputs_map, key, False) for key in required_inputs]
+            optional_inputs = ["axes", "steps"]
+            axes, steps = [self._process_slice_input(inputs_map, key, True) for key in optional_inputs]
+            inputs = [data, starts, ends]
+            if axes is None:
+                if steps is not None:
+                    inputs.extend("", steps)
+            else:
+                inputs.append(axes)
+                if steps is not None:
+                    inputs.append(steps)
```

Comment:
inputs = [data, starts, ends, axes, steps]
then remove trailing None, then replace None to ""